### PR TITLE
bpf: nodeport: only add DSR Extension to TCPv6 SYN packet

### LINKED ISSUE #21991: CFP: Limit DSR via IPv6 Extension Header to TCP-SYN
## Cilium Feature Proposal

**Describe the feature you'd like**

The IPv4 LB code [implements](https://github.com/cilium/cilium/blob/4b78c1e60d113fc081cdecffa1a9617bcd44cfd2/bpf/lib/nodeport.h#L1306) an optimization, where it adds the DSR IP Option only to TCP-SYN packets.

Looks like we didn't add this to the IPv6 [path](https://github.com/cilium/cilium/blob/4b78c1e60d113fc081cdecffa1a9617bcd44cfd2/bpf/lib/nodeport.h#L238) of the feature.

Would be nice to add a few unit tests for this feature while at it, to validate that the resulting packets look as expected.

### PR DESCRIPTION
**Pull request by julianwiedmann**
For DSR on TCP connections, we only want to insert the DSR information into
the SYN packet. This is an obvious optimization, and allows us to avoid any
risk of exceeding the MTU.

The support for this functionality currently only exists in the IPv4 path.
Also add it to the IPv6 path.

Fixes: https://github.com/cilium/cilium/issues/21991

```release-note
Reduce the risk of packet fragmentation on IPv6 when using KPR with DSR mode.
```

### CODE REVIEW

**Comment by julianwiedmann** [comment]
/test-1.26-net-next

**Comment by julianwiedmann** [comment]
Note that this might not strictly qualify as important fix to backport. But we'll need it as prep-work for https://github.com/cilium/cilium/pull/22978.

**Review by YutaroHayakawa — APPROVED** [review]

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -222,10 +222,27 @@ static __always_inline int dsr_set_ext6(struct __ctx_buff *ctx,
...
 	/* The IPv6 extension should be 8-bytes aligned */
 	build_bug_on((sizeof(struct dsr_opt_v6) % 8) != 0);
 
+	hdrlen = ipv6_hdrlen(ctx, &nexthdr);
+	if (hdrlen < 0)
+		return hdrlen;
+
+	/* See dsr_set_opt4(): */
+	if (nexthdr == IPPROTO_TCP) {
+		union tcp_flags tcp_flags = { .value = 0 };
```
Comment (aspsk) [inline bpf/lib/nodeport.h:237]: Why not just `= {}`? :-)

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -222,10 +222,27 @@ static __always_inline int dsr_set_ext6(struct __ctx_buff *ctx,
...
+
+		if (l4_load_tcp_flags(ctx, ETH_HLEN + hdrlen, &tcp_flags) < 0)
+			return DROP_CT_INVALID_HDR;
+
+		if (!(tcp_flags.value & (TCP_FLAG_SYN)))
```
Comment (aspsk) [inline bpf/lib/nodeport.h:242]: `value & TCP_FLAG_SYN`? (no braces)

**Review by aspsk — APPROVED** [review]
Looks great, only two cosmetic comments.

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:237]: copy&pasted from the IPv4 path :smile:. I'll probably be adding this pattern a few more times in another PR, will try to bake a clean helper (`nodeport_dsr_needs_info()` ?) for it then.

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:242]: see above

**Review by nebril — APPROVED** [review]
ci bits LGTM

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
The release note could probably be reworded to:
```
When using KPR NodePort with DSR mode for IPv6, only add the DSR Extension to SYN packets.
```
To avoid saying that only TCPv6 SYN packets receive the DSR extension :grin:

**Comment by pchaigno** [comment]
Hm, actually, we should describe the impact for the user in the release note, not the change itself. What's the impact here? Do we have any regressions w.r.t. MTU or is it fixing potential issues? Maybe:
```
Reduce the risk of packet fragmentation on IPv6 when using KPR with DSR mode.
```

**Comment by julianwiedmann** [comment]
Runtime hit flake https://github.com/cilium/cilium/issues/21999 for `pkg/crypto/certloader` in `TestWatchedServerConfigRotation`

**Comment by julianwiedmann** [comment]
> Hm, actually, we should describe the impact for the user in the release note, not the change itself. What's the impact here? Do we have any regressions w.r.t. MTU or is it fixing potential issues?

At this stage it's fixing potential issues, not sure how much use the IPv6 part is getting in the field.


> 
> ```
> Reduce the risk of packet fragmentation on IPv6 when using KPR with DSR mode.
> ```

Shamelessly stole that :+1:.

**Pull request merged**